File: job-offer/src/main/java/com/oxygensend/joboffer/infrastructure/redis/CacheConfiguration.java
package com.oxygensend.joboffer.infrastructure.redis;

import java.time.Duration;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.data.redis.cache.RedisCacheConfiguration;
import org.springframework.data.redis.cache.RedisCacheManager;
import org.springframework.data.redis.connection.RedisConnectionFactory;

@Configuration
public class CacheConfiguration {


    @Bean
    public RedisCacheManager cacheManager(RedisConnectionFactory redisConnectionFactory) {
        return RedisCacheManager.RedisCacheManagerBuilder
                .fromConnectionFactory(redisConnectionFactory)
                .cacheDefaults(redisCacheConfiguration())
                .build();
    }

    @Bean
    public RedisCacheConfiguration redisCacheConfiguration() {
        return RedisCacheConfiguration.defaultCacheConfig()
                                      .entryTtl(Duration.ofMinutes(180))
                                      .disableCachingNullValues();
    }
}

File: static-data/src/main/kotlin/com/oxygensend/staticdata/infrastructure/caffeine/CacheConfiguration.kt
package com.oxygensend.staticdata.infrastructure.caffeine

import com.github.benmanes.caffeine.cache.Caffeine
import com.oxygensend.staticdata.infrastructure.StaticDataProperties
import org.springframework.cache.CacheManager
import org.springframework.cache.caffeine.CaffeineCacheManager
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration

@Configuration
class CacheConfiguration(private val staticDataProperties: StaticDataProperties) {

    @Bean
    fun statCacheConfig(): Caffeine<Any, Any> = Caffeine.newBuilder()
        .expireAfterWrite(staticDataProperties.cache!!.statCacheTtl)


    @Bean
    fun caffeineStatCacheManager(): CacheManager = CaffeineCacheManager().apply { setCaffeine(statCacheConfig()) }

}